awswrangler/s3/_read_orc.py

```diff
@@ -0,0 +1,651 @@
+        nullable dtypes are used for all dtypes that have a nullable implementation when
+        “numpy_nullable” is set, pyarrow is used for all dtypes if “pyarrow” is set.
+
+        The dtype_backends are still experimential. The "pyarrow" backend is only supported with Pandas 2.0 or above.
+    use_threads : Union[bool, int], default True
+        True to enable concurrent requests, False to disable multiple threads.
+        If enabled, os.cpu_count() is used as the max number of threads.
+        If integer is provided, specified number is used.
+    ray_args: typing.RaySettings, optional
+        Parameters of the Ray Modin settings. Only used when distributed computing is used with Ray and Modin installed.
+    boto3_session : boto3.Session(), optional
+        Boto3 Session. The default boto3 session is used if None is received.
+    s3_additional_kwargs : Optional[Dict[str, Any]]
+        Forward to S3 botocore requests.
+    pyarrow_additional_kwargs : Dict[str, Any], optional
+        Forwarded to `to_pandas` method converting from PyArrow tables to Pandas DataFrame.
+        Valid values include "split_blocks", "self_destruct", "ignore_metadata".
+        e.g. pyarrow_additional_kwargs={'split_blocks': True}.
+
+    Returns
+    -------
+    pandas.DataFrame
+        Pandas DataFrame.
+
+    Examples
+    --------
+    Reading all ORC files under a prefix
+
+    >>> import awswrangler as wr
+    >>> df = wr.s3.read_orc(path='s3://bucket/prefix/')
+
+    Reading all ORC files from a list
+
+    >>> import awswrangler as wr
+    >>> df = wr.s3.read_orc(path=['s3://bucket/filename0.orc', 's3://bucket/filename1.orc'])
+
+    Reading ORC Dataset with PUSH-DOWN filter over partitions
+
+    >>> import awswrangler as wr
+    >>> my_filter = lambda x: True if x["city"].startswith("new") else False
+    >>> df = wr.s3.read_orc(path, dataset=True, partition_filter=my_filter)
+
+    """
+    ray_args = ray_args if ray_args else {}
+
+    s3_client = _utils.client(service_name="s3", session=boto3_session)
+    paths: List[str] = _path2list(
+        path=path,
+        s3_client=s3_client,
+        suffix=path_suffix,
+        ignore_suffix=_get_path_ignore_suffix(path_ignore_suffix=path_ignore_suffix),
+        last_modified_begin=last_modified_begin,
+        last_modified_end=last_modified_end,
+        ignore_empty=ignore_empty,
+        s3_additional_kwargs=s3_additional_kwargs,
+    )
+    if not path_root:
+        path_root = _get_path_root(path=path, dataset=dataset)
+    if path_root and partition_filter:
+        paths = _apply_partition_filter(path_root=path_root, paths=paths, filter_func=partition_filter)
+    if len(paths) < 1:
+        raise exceptions.NoFilesFound(f"No files Found on: {path}.")
+
+    version_ids = _check_version_id(paths=paths, version_id=version_id)
+
+    # Create PyArrow schema based on file metadata, columns filter, and partitions
+    schema: Optional[pa.schema] = None
+    if validate_schema:
```

Comment:
This schema validation code also looks 99% the same between parquet and orc. Shall we move it to a method that accepts a reader?